hibike/devices/ExampleDevice/example_device.cpp

```diff
@@ -1,69 +1,138 @@
 
+// each device is responsible for keeping track of it's own params
 uint8_t value0;
 uint8_t value1;
 uint16_t value2;
 uint32_t value3;
 uint64_t value4;
+uint64_t value5;
+
 
 // normal arduino setup function, you must call hibike_setup() here
 void setup() {
   hibike_setup();
 }
 
+
 // normal arduino loop function, you must call hibike_loop() here
 // hibike_loop will look for any packets in the serial buffer and handle them
-void loop() {
 
+void loop() {
   // do whatever you want here
   // note that hibike will only process one packet per call to hibike_loop()
   // so exessive delays here will affect hibike.
-  value1++;
-  value2++;
-  value3++;
-  value4++;
+  value1 += 1;
+  value2 += 2;
+  value3 += 3;
+  value4 += 4;
+  value5 += 5;
 
   hibike_loop();
 }
 
 
-// you must implement this function. It is called when the device receives a DeviceUpdate packet.
-// the return value is the value field of the DeviceRespond packet hibike will respond with
-uint32_t device_update(uint8_t param, uint32_t value) {
-  if (param < NUM_PARAMS) {
-    params[param] = value;
-    return params[param];
+// You must implement this function.
+// It is called when the device receives a Device Write packet.
+// Updates param to new value passed in data.
+//    param   -   Parameter index
+//    data    -   value to write, in bytes TODO: What endian?
+//    len     -   number of bytes in data
+//
+///   return  -   size of bytes written on success; otherwise return 0
+
+uint32_t device_write(uint8_t param, uint8_t* data, size_t len) {
+  switch (param) {
+    case VALUE_0:
+      value0 = (uint8_t) read_num_bytes(data, len);
+      return sizeof(value0);
+      break;
+    case VALUE_1:
+      value1 = (uint8_t) read_num_bytes(data, len);
+      return sizeof(value1);
+      break;
+    case VALUE_2:
+      value2 = (uint16_t) read_num_bytes(data, len);
+      return sizeof(value2);
+      break;
+    case VALUE_3:
+      value3 = (uint32_t) read_num_bytes(data, len);
+      return sizeof(value3);
+      break;
+    case VALUE_4:
+      value4 = (uint64_t) read_num_bytes(data, len);
+      return sizeof(value4);
+      break;
+    case VALUE_5:
+      value5 = (uint64_t) read_num_bytes(data, len);
+      return sizeof(value5);
+      break;
+    default:
+      return 0;
   }
-  return ~((uint32_t) 0);
+  return 0;
 }
 
-// you must implement this function. It is called when the devie receives a DeviceStatus packet.
-// the return value is the value field of the DeviceRespond packet hibike will respond with
-uint32_t device_status(uint8_t param) {
-  if (param < NUM_PARAMS) {
-    return params[param];
+
+// You must implement this function.
+// It is called when the device receives a Device Data Update packet.
+// Modifies data_update_buf to contain the parameter value.
+//    param           -   Parameter index
+//    data_update_buf -   buffer to return data in
+//    buf_len         -   Maximum length of the buffer? TODO: Clarify
```

Comment:
buf_len is the number of bytes left in the buffer, so this function only writes to the buffer if their is sufficient space left